app/backend/src/couchers/servicers/auth.py

```diff
@@ -424,27 +424,48 @@ def CompletePasswordReset(self, request, context):
             else:
                 context.abort(grpc.StatusCode.NOT_FOUND, errors.INVALID_TOKEN)
 
-    def CompleteChangeEmail(self, request, context):
-        """
-        Completes an email change request.
-
-        Removes the old email and replaces with the new
-        """
+    def ConfirmChangeEmail(self, request, context):
         with session_scope() as session:
-            user = (
+            user_with_valid_token_from_old_email = (
+                session.query(User)
+                .filter(User.old_email_token == request.change_email_token)
+                .filter(User.old_email_token_created <= func.now())
+                .filter(User.old_email_token_expiry >= func.now())
+                .one_or_none()
+            )
+            user_with_valid_token_from_new_email = (
                 session.query(User)
                 .filter(User.new_email_token == request.change_email_token)
                 .filter(User.new_email_token_created <= func.now())
                 .filter(User.new_email_token_expiry >= func.now())
                 .one_or_none()
             )
-            if user:
-                user.email = user.new_email
-                user.new_email = None
+
+            if user_with_valid_token_from_old_email:
+                user = user_with_valid_token_from_old_email
+                user.old_email_token = None
+                user.old_email_token_created = None
+                user.old_email_token_expiry = None
+                user.confirmed_email_change_via_old_email = True
+            elif user_with_valid_token_from_new_email:
+                user = user_with_valid_token_from_new_email
                 user.new_email_token = None
                 user.new_email_token_created = None
                 user.new_email_token_expiry = None
-                session.commit()
-                return empty_pb2.Empty()
+                user.confirmed_email_change_via_new_email = True
             else:
                 context.abort(grpc.StatusCode.NOT_FOUND, errors.INVALID_TOKEN)
+
+            if user.confirmed_email_change_via_old_email and user.confirmed_email_change_via_new_email:
+                user.email = user.new_email
+                user.new_email = None
+                user.confirmed_email_change_via_old_email = False
+                user.confirmed_email_change_via_new_email = False
+                res = auth_pb2.ConfirmChangeEmailRes(success=True)
+            elif user.confirmed_email_change_via_old_email:
+                res = auth_pb2.ConfirmChangeEmailRes(requires_confirmation_from_new_email=True)
+            else:
+                res = auth_pb2.ConfirmChangeEmailRes(requires_confirmation_from_old_email=True)
+
+            session.commit()
```

Comment:
No need for commit here (when execution leaves the ctx manager; it'll automatically commit, and if it fails, it'll error out regardless)

Reply:
Can you explain this to me in more detail sometime? I still can't figure out when to commit and when not